manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+	//     if we get to this point, then we have been through every desired
+	//     network ID and not found one matching the one on this VIP, so we
+	//     can add it to the list of VIPs to deallocate
+vips:
+	for _, vip := range endpoint.VirtualIPs {
+		for _, nwid := range networkIDs {
+			if vip.NetworkID == nwid {
+				keep = append(keep, vip)
+				continue vips
+			}
+		}
+		deallocate = append(deallocate, vip)
+	}
+
+	// now figure out which new network IDs we've added, which is the same loop
+	// above but swapped around to check network IDs against VIPs
+newvips:
+	for _, nwid := range networkIDs {
+		for _, vip := range endpoint.VirtualIPs {
+			if vip.NetworkID == nwid {
+				continue newvips
+			}
+		}
+		allocate = append(allocate, nwid)
+	}
+
+	// now deallocate the old vips, and allocate the new ones
+	a.deallocateVIPs(deallocate)
+
+	// create a new slice to hold the vips we're allocating now
+	newVips := make([]*api.Endpoint_VirtualIP, 0, len(allocate))
+allocateLoop:
+	for _, nwid := range allocate {
+		// we already verified that every one of the requested networks
+		// existed, so again, no need to check ok
+		local := a.networks[nwid]
+		// we don't need to nil check any of the intermediate fields. we made
+		// them we we know they're filled in
+		ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+		opts := local.nw.IPAM.Driver.Options
+		// the network may have several pools of IP addresses, and some of them
+		// may be full, so we need to try allocation on every pool available
+		// for the network until we find a pool that succeeds.
+		for _, poolID := range local.pools {
+			// passing nil for the second args indicates that we don't have any
+			// particular address in mind to allocate
+			ip, _, err := ipam.RequestAddress(poolID, nil, opts)
+			// if there's no error, we have a valid address and we're done
+			if err == nil {
+				// add it to the endpoints map so we can figure out what pool
+				// it belongs to when we deallocate
+				local.endpoints[ip.String()] = poolID
+				// add a new VIP object to our slice
+				newVips = append(newVips, &api.Endpoint_VirtualIP{
+					NetworkID: nwid,
+					// ip.String() will return an IP in CIDR notation. that's
+					// intended, current behavior is for vips to be CIDR
+					// notation.
+					Addr: ip.String(),
+				})
+				// continue allocate loop, to skip the error handling that
+				// occurs when ever pool has been exhausted
+				continue allocateLoop
+			}
+			// if we get ErrNoAvailableIPs, it means this pool is already full,
+			// and we'll go on to the next one. if we get any other error,
+			// that's fatal, so we'll return it and clean up
+			if err != ipamapi.ErrNoAvailableIPs {
+				return ErrFailedAddressRequest{"new", err}
+			}
+		}
+		// if we get here, that means we've tried every pool on the network,
+		// and all of them are exhausted
+		return ErrFailedAddressRequest{"new", ipamapi.ErrNoAvailableIPs}
```

Comment:
Apologies, not super familiar with what is important to allocate and deallocate, so this may not make sense.  Do we need to have rollback logic for any addresses or VIPS that have already been allocated if this error or `ErrFailedAddressRequest` is returned partway through the loop?

Reply:
I forgot to add rollback logic, but I have it added in my local branch.